### ISSUE TITLE
Externally scoped var wrongfully accused of no-unused-vars

### ISSUE THREAD

**Issue by Redsandro** [eslint/eslint#5942]
**What version of ESLint are you using?**

Using `eslint` 1.5.0

**What parser (default, Babel-ESLint, etc.) are you using?**

[`linter-eslint`](https://github.com/AtomLinter/linter-eslint) 7.2.0 on Atom

**Please show your full configuration:**

[Default](https://github.com/eslint/eslint/blob/master/conf/eslint.json)

**What did you do? Please include the actual source code causing the issue.**

This script is legal, but ESLint has trouble detecting the use of `number` and throws `no-unused-vars`

```javascript
'use strict';

var number; // no-unused-vars

new Promise((res, rej) => {res(43);}).then(count => {
	number = count;
});
```

**What did you expect to happen?**

No messages from ESLint.

**What actually happened?**

See comment in above code. When commenting out the allegedly unused var `number`, a new warning pops up claiming `number` is not defined:

```javascript
'use strict';

// var number;

new Promise((res, rej) => {res(43);}).then(count => {
	number = count; // no-undef
});
```

**Please include the actual, raw output from ESLint.**

Not sure how, Atom uses ESLint internally. I assume mentioning the codes I got from Atom is clear, because the problem isn't very complex.

**Comment by eslintbot** [comment]
Hi @Redsandro, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Hi @Redsandro! If I understood your question correctly, you want to mark `number` as used since other scripts will read it, correct?

If so, you might want to take a look at the `no-unused-var` docs, specifically the section about the [`/* exported */ comment](http://eslint.org/docs/rules/no-unused-vars#exported). If you use those comments, that should be enough to satisfy the `no-unused-vars` rule.

**Comment by Redsandro** [comment]
> you want to mark number as used since other scripts will read it, correct?

No. The code is a single script. It is legal `var` usage to define `number` and have it used in an anonymous function within the same scope, as provided by the promise. In the second piece of code, you can see that ESLint expects `count` to be defined. However, when I define it as in the first piece of code, ESLint thinks `count` is never used, when it clearly is.

The `no-undef` code in the second script is correct. The `no-unused-vars` in the first script is incorrect, because nothing is exported or imported. A `var` is defined and used in a function within it's scope.

> please (..) fill out the bug report template

Sorry, I assumed I answered all questions implicitly. I edited my post. :smiley:

**Comment by nzakas** [comment]
Confirmed this in the demo..

**Comment by ilyavolodin** [comment]
I can only see number being assigned to, but nothing is actually consuming the value of number in your code, right? Or am I missing something?

**Comment by michaelficarra** [comment]
This is not a bug.

**Comment by mysticatea** [comment]
```js
'use strict';

var number; // no-unused-vars

new Promise((res, rej) => {res(43);}).then(count => {
    number = count;
});
```

`number` and `rej` in this code are unused variables. It's an intentional behavior.

**Comment by platinumazure** [comment]
@Redsandro As pointed out by others, the rule is designed to flag situations where variables are only assigned to. Nowhere in your code is `number` read, so it is considered "unused" for the purpose of `no-unused-vars`. I mentioned `/* exported */` in case this was a shared global situation, but it seems that is not the case.

At this point, with the information we have, I don't think this is a bug and I think the rule is correctly saying you aren't using the variable.

**Comment by kaicataldo** [comment]
From the [no-unused-vars rule docs](http://eslint.org/docs/rules/no-unused-vars):
> A variable is not considered to be used if it is only ever assigned to (var x = 5) or declared.

**Comment by Redsandro** [comment]
Shoot, you are all correct. `number` is never used. This was a made up example contrived from a bigger work where I experienced this problem. I was sure the Linter made a mistake, and I had somewhat of a tunnel vision since. But the shameful truth is that I made a mistake and had recently deprecated the use of the variable.

It's with regret that I have to inform you that I've been wasting your time. Thanks for looking into this though!

**Comment by nzakas** [comment]
Doh, I should know better than to triage bugs when I'm overly tired. :( Thanks for the save, everyone!

**Comment by kaicataldo** [comment]
@Redsandro No worries - glad we got to the bottom of it 😄
